File: lib/otel/resource.go
package otel

import (
	"maps"
	"strings"

	"github.com/AtomiCloud/diene.go-interfaces/lib/interfaces"
	semconv "go.opentelemetry.io/otel/semconv/v1.41.0"
)

// Semantic-convention resource attribute keys the service tree maps onto
// (C0 §4). They are taken from the SDK's semconv package rather than hand-typed,
// so a convention rename cannot silently change the wire.
var (
	// AttrDeploymentEnvironmentName carries the landscape.
	AttrDeploymentEnvironmentName = string(semconv.DeploymentEnvironmentNameKey)
	// AttrServiceNamespace carries the platform.
	AttrServiceNamespace = string(semconv.ServiceNamespaceKey)
	// AttrServiceName carries the service.
	AttrServiceName = string(semconv.ServiceNameKey)
	// AttrServiceVersion carries the version.
	AttrServiceVersion = string(semconv.ServiceVersionKey)
)

// Raw AtomiCloud taxonomy attribute keys. They ship ALONGSIDE the semantic
// conventions so queries can address the unmapped coordinate (module) and the
// untranslated taxonomy directly.
const (
	// AttrAtomiLandscape is the raw landscape attribute.
	AttrAtomiLandscape = "atomi.landscape"
	// AttrAtomiPlatform is the raw platform attribute.
	AttrAtomiPlatform = "atomi.platform"
	// AttrAtomiService is the raw service attribute.
	AttrAtomiService = "atomi.service"
	// AttrAtomiModule is the raw module attribute.
	AttrAtomiModule = "atomi.module"
	// AttrAtomiVersion is the raw version attribute.
	AttrAtomiVersion = "atomi.version"
)

// Standard OTEL_* resource environment variables honored as an ops escape hatch.
const (
	// EnvResourceAttributes overrides or extends resource attributes.
	EnvResourceAttributes = "OTEL_RESOURCE_ATTRIBUTES"
	// EnvServiceName overrides service.name.
	EnvServiceName = "OTEL_SERVICE_NAME"
)

// ResourceAttributes maps a service-tree identity onto the canonical resource
// attribute set: the four C0 §4 semantic conventions plus the five raw
// `atomi.*` taxonomy attributes.
func ResourceAttributes(identity AppIdentity) (map[string]string, error) {
	if identityErr := identity.Validate(); identityErr != nil {
		return nil, identityErr
	}
	app := identity.Trimmed()
	return map[string]string{
		AttrDeploymentEnvironmentName: app.Landscape,
		AttrServiceNamespace:          app.Platform,
		AttrServiceName:               app.Service,
		AttrServiceVersion:            app.Version,
		AttrAtomiLandscape:            app.Landscape,
		AttrAtomiPlatform:             app.Platform,
		AttrAtomiService:              app.Service,
		AttrAtomiModule:               app.Module,
		AttrAtomiVersion:              app.Version,
	}, nil
}

// ParseResourceAttributes parses an OTEL_RESOURCE_ATTRIBUTES value in W3C
// baggage-like `key=value,key2=value2` form. Entries without a key are skipped
// rather than failing the process: an ops escape hatch must never be able to
// crash a service through a typo.
func ParseResourceAttributes(value string) map[string]string {
	attributes := map[string]string{}
	for entry := range strings.SplitSeq(value, ",") {
		trimmed := strings.TrimSpace(entry)
		if trimmed == "" {
			continue
		}
		name, attributeValue, found := strings.Cut(trimmed, "=")
		name = strings.TrimSpace(name)
		if !found || name == "" {
			continue
		}
		attributes[name] = strings.TrimSpace(attributeValue)
	}
	return attributes
}

// ResolvedResourceAttributes applies the standard OTEL_* overrides on top of the
// configured identity mapping: OTEL_RESOURCE_ATTRIBUTES entries win over the
// derived attributes, and OTEL_SERVICE_NAME wins over service.name. This is the
// C0 §4 ops escape hatch — the block never clobbers a set OTEL_* variable.
func ResolvedResourceAttributes(identity AppIdentity, system interfaces.System) (map[string]string, error) {
	configured, configuredErr := ResourceAttributes(identity)
	if configuredErr != nil {
		return nil, configuredErr
	}
	rawAttributes, rawErr := EnvValue(system, EnvResourceAttributes)
	if rawErr != nil {
		return nil, rawErr
	}
	if rawAttributes != nil {
		maps.Copy(configured, ParseResourceAttributes(*rawAttributes))
	}
	serviceName, serviceErr := EnvValue(system, EnvServiceName)
	if serviceErr != nil {
		return nil, serviceErr
	}
	if serviceName != nil && strings.TrimSpace(*serviceName) != "" {
		configured[AttrServiceName] = strings.TrimSpace(*serviceName)
	}
	return configured, nil
}

File: src/lib/resource.ts
import type { TelemetryAttributes } from '@atomicloud/diene.interfaces';
import { resourceFromAttributes, type Resource } from '@opentelemetry/resources';
import {
  ATTR_DEPLOYMENT_ENVIRONMENT_NAME,
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_NAMESPACE,
  ATTR_SERVICE_VERSION,
} from '@opentelemetry/semantic-conventions';
import { z } from 'zod';

const identityValueSchema = z.string().trim().min(1);

const appIdentitySchema = z
  .object({
    landscape: identityValueSchema,
    platform: identityValueSchema,
    service: identityValueSchema,
    module: identityValueSchema,
    version: identityValueSchema,
  })
  .strict()
  .readonly();

type AppIdentity = z.infer<typeof appIdentitySchema>;
type OtelEnvironment = Readonly<Record<string, string | undefined>>;

function mapResourceAttributes(identity: AppIdentity): TelemetryAttributes {
  const app = appIdentitySchema.parse(identity);
  return Object.freeze({
    [ATTR_DEPLOYMENT_ENVIRONMENT_NAME]: app.landscape,
    [ATTR_SERVICE_NAMESPACE]: app.platform,
    [ATTR_SERVICE_NAME]: app.service,
    [ATTR_SERVICE_VERSION]: app.version,
    'atomi.landscape': app.landscape,
    'atomi.module': app.module,
    'atomi.platform': app.platform,
    'atomi.service': app.service,
    'atomi.version': app.version,
  });
}

function parseOtelResourceAttributes(value: string | undefined): TelemetryAttributes {
  if (value === undefined || value.trim() === '') return Object.freeze({});
  const entries = value
    .split(',')
    .map(entry => entry.trim())
    .filter(entry => entry.length > 0)
    .map(entry => {
      const separator = entry.indexOf('=');
      if (separator <= 0) return undefined;
      const key = entry.slice(0, separator).trim();
      const attributeValue = entry.slice(separator + 1).trim();
      return key.length === 0 ? undefined : ([key, attributeValue] as const);
    })
    .filter((entry): entry is readonly [string, string] => entry !== undefined);
  return Object.freeze(Object.fromEntries(entries));
}

function resourceAttributes(identity: AppIdentity, environment: OtelEnvironment = process.env): TelemetryAttributes {
  const configured = mapResourceAttributes(identity);
  const standard = parseOtelResourceAttributes(environment.OTEL_RESOURCE_ATTRIBUTES);
  return Object.freeze({
    ...configured,
    ...standard,
    ...(environment.OTEL_SERVICE_NAME === undefined || environment.OTEL_SERVICE_NAME.trim() === ''
      ? {}
      : { [ATTR_SERVICE_NAME]: environment.OTEL_SERVICE_NAME.trim() }),
  });
}

function createOtelResource(identity: AppIdentity, environment: OtelEnvironment = process.env): Resource {
  return resourceFromAttributes(resourceAttributes(identity, environment));
}

export type { AppIdentity, OtelEnvironment };
export {
  appIdentitySchema,
  createOtelResource,
  mapResourceAttributes,
  parseOtelResourceAttributes,
  resourceAttributes,
};
